### DISCUSSION TITLE
Next.js for internal apps

### DISCUSSION THREAD

**Discussion by nareshbhatia** [vercel/next.js#44703] [Help] [answered]
I am trying to make a case for using Next.js for internal apps, e.g. dashboards used by ops teams or workflow tools used by customer service representatives. The primary characteristics here are:

- We have captive audience that uses the app all day, so metrics like FCP and TTI don't matter.
- Users are mostly on a desktop and a fast network, so apps can be moderately heavy (within reason)

Trying to get a sense for using Next.js for this use case vs. a SPA using simply React and CSR. Any experiences or opinions on either approach?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Well, Next.js is an SPA - and I am assuming that when you say "React and CSR", you mean one of the CRA, Vite, options out there, in full CSR mode.

When we say CSR often we mean a static HTML site, that's just a shell awaiting for a JS entry point, the main bundle, which is in charge of populating the page with HTML + CSS, and interactivity.

Next.js can do all of that. 

An SPA is an application that uses the history API, and through a modified anchor tag, prevents the browser's default behaviour of unloading the page and navigating elsewhere. Instead, the application's router takes over and renders content we have defined for the new url, which mimics the browser behaviour, while also avoiding blank pages, resource refetching, etc, etc...

Next.js does that.

Next.js works primarily in server mode, through `next start`, but can also do `next export` where a static HTML site is "exported". 

Next.js has a bunch of utility functions to get data. GetStaticProps, GetServerSideProps, GetStaticPaths, and API routes. Middleware is also available nowadays, and when deployed without edge support, it simply runs next to your next server, so no special handling is needed.

Next.js has also a file system based routing solution, which means that files places under `pages` are the routes your application exposes, so there's no need to bring in a 3rd party router into the mix. 

Since API routes are a built-in backend for your frontend, you can hide API URLS, secrets, provide authorisation and authentication, all within the same app - a CSR app would need to have this values included in the main bundle, or would need a backend as a jump-pad to external resources.

Configuring Next.js is pretty simple too, and it does a good job at keeping Webpack under control, for the most part ~ 

There's also code elimination: https://next-code-elimination.vercel.app/ 

I feel like I should point out some cons: 

- I'd say that one thing against using Next.js for your specific use case, is that it really breaks down the JavaScript into a very optimised, minified, compatible with most browsers version. The latest releases have dropped IE, but still the code might look ugly, which might or might not matter for an app that sits on an internal circuit, where one would expect people use modern browsers with high speed internet. Take this with a grain of salt, because, one could also say that it doesn't matter at all.
- Trailing slashes in exported mode can be a bit of a difficult bump.
- The team releases very fast, which is a good thing, but you might feel like you need to update the next.js version very often - FOMO
- Hydration issues are a common pitfalls of moving from CSR to SSR
- Memory leaks and what not, if running a dedicated server, I'd say use Vercel if possible
- There's no DOM/Window during SSR and this is not intuitive to people used to full 100% CSR
- Often the idea of having a server which shares JavaScript contexts with many users, is not intuitive to devs used to CSR (from my own experience working with others) 

And that's it - there's nothing really other than that, you just need to learn to use Next.js - and get a sense of why GetServerSideProps is such an OP, albeit prone to being slow, model.

> Any experiences or opinions on either approach?

I think given that you can just grab a boilerplate, like https://github.com/ixartz/Next-js-Boilerplate  or start with create-next-app, and get ready to go within minutes, and not have to setup routers, proxies, or any of that jazz to work with 3rd party data - is a pretty good argument to take Next.js

I've used in a few internal app projects, and it has always resulted in happy development. Usually the issues are on the side of setting up CDN's or caching for certain pages, since people often want to separate concerns and keep Next.js as simply a rendering engine. Rarely have I seen a use case where the file system is a problem or where something is just plain impossible to do because of the  model.

**Reply by nareshbhatia** [reply]
@icyJoseph, thank you so much for a detailed response. I really appreciate it.

You are correct in your assumption that by "SPA using simply React and CSR" I was referring to apps generated by tools such as CRA and Vite. It is interesting that you mentioned "Next.js is an SPA". I think of it primarily as a multipage app because each page is a separate entry point that gets independently downloaded. For me, SPAs are applications that are downloaded in one shot and from there on all decisions such as routing and fetching are made by the client. What was interesting to me is that Next.js can act like a SPA using `next export`. Articles like [this](https://colinhacks.com/essays/building-a-spa-with-nextjs) and [this](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/) (though dated) made me believe that this is incredibly hard. So thank you for pointing this out.  Will have to play with `next export`.

While I am totally convinced of the power of Next.js (or OP as you mentioned :blush:), I was trying to build a justification for internal apps where teams are stuck on the simplicity of CRA/vite apps and the ease with which they can be deployed statically. Also there is the inertia of moving to a file-based router instead of the well-known react-router. So the justification has to go over and above "Next.js can do it too". Your reply has brought out several such points, notably:

1. Ability to hide API URLs & secrets and provide authorization and authentication all within the same app
2. Much better developer experience, especially when deployed on Vercel infrastructure

I would also add the following:

1. Automatic code splitting, page preloading & dynamic imports, which help scaling of large internal apps
2. Ability to fetch data and render it server-side, which is faster than fetching on the client side (i.e. SSR)
3. Ability to mix client-side and server-side components (Next.js 13 & RSC), combining the rich interactivity of client-side apps with the improved performance of traditional server rendering

Anything I have missed?

**Reply by icyJoseph** [reply]
Nope, you got it. The code splitting is a good one, forgot to mention it with code elimination. 

The point number two is what I meant with data fetching functions. 

And yeah, pages that are pre-rendered do generate their standalone entry point to the app. 

However GSSP pages are rendered on demand, in two ways. On SPA transitions GSSP runs server side and the response is passed as props to the page component. And upon arrival to the app a whole HTML page is generated server side, by calling GSSP and doing SSR.

**Reply by nareshbhatia** [reply]
Thank you for the clarifications, @icyJoseph. I think we have covered all bases.

Welcome opinions from others!

**Reply by GabenGar** [reply]
> Next.js has also a file system based routing solution, which means that files places under pages are the routes your application exposes, so there's no need to bring in a 3rd party router into the mix.

Not entirely true for the static export, since it will build the dynamic parts with all the nextjs-specific syntax of brackets, if the path wasn't consumed by `getStaticProps()`. And that means you'll have to write an ad-hoc script which will create a mapping between nextjs paths and whatever path syntax your static server uses. Otherwise the paths like `/entity/[id]` will work client-side but fail when accessed directly from browser url bar.

**Reply by nareshbhatia** [reply]
Good to know, @GabenGar.

**Comment by laurens-runnable** [comment]
In my experience the three biggest benefits of Next.js compared to vanilla React are:

* SSR
* Routing
* Backend (i.e. API routes)

SSR doesn't give you much benefit if your users keep the app open most of the day. Even less so when they're on a fast network.

Routing requirements are already met, since your team is already using react-router.

That leaves backend. One example of backend functionality is SSO, which is typically implemented with OpenID Connect. Do bear in mind that there are substantial differences between the `nodejs` and `edge` runtimes.

**Reply by nareshbhatia** [reply]
Thank you for your opinions, @laurens-runnable. Can you please explain the last point regarding SSO in more detail? What are the benefits of using Next.js in this case and what are the difference between `nodejs` and `edge` runtimes?

**Reply by nareshbhatia** [reply]
@laurens-runnable, also what do you think about automatic code splitting, page preloading & dynamic imports? Do you think these features will help in internal apps? While some of these can be done manually, having them done automatically would be great DX.

**Reply by laurens-runnable** [reply]
> Thank you for your opinions, @laurens-runnable. Can you please explain the last point regarding SSO in more detail? What are the benefits of using Next.js in this case and what are the difference between `nodejs` and `edge` runtimes?

Sure @nareshbhatia, glad to be of help.

I mentioned SSO as an example, because 1) it's a common requirement for internal apps and 2) it requires a backend due to the use of client IDs and client secrets. The benefit of using Next.js here is simply that you don't need another backend, at least for the purpose of SSO, potentially reducing the number of deployment artifacts.

I'm not too familiar with the `edge` runtime, since it's quite new, but I do know it has quite a few limitations compared to the `nodejs` runtime. These limitations show up in middleware, which only supports the `edge` runtime right now.

See this discussion for more info:
https://github.com/vercel/next.js/discussions/34179

**Reply by laurens-runnable** [reply]
> @laurens-runnable, also what do you think about automatic code splitting, page preloading & dynamic imports? Do you think these features will help in internal apps? While some of these can be done manually, having them done automatically would be great DX.

I don't think any of these features will have much benefit, to be honest. These features are all geared towards reducing initial page load, which is not much of an issue for internal apps to begin with.

**Reply by nareshbhatia** [reply]
Thank you, @laurens-runnable. Your comments are very helpful in forming my own opinions.
